docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md

```diff
@@ -0,0 +1,111 @@
+|-----------------|-------------------|---------------------------------------------------------------------------------------|---------|----------|
+| `keys`          | `array(string)`   | Keys that will be used to group the spans, log records, or metric data points together. |         | no       |
+| `output`        | [output][]        | Configures where to send received telemetry data.                                     | yes     |          |
+| `debug_metrics` | [debug_metrics][] | Configures the metrics that this component generates to monitor its state.            | no      |          |
+
+[output]: #output-block
+[debug_metrics]: #debug_metrics-block
+
+### keys
+`keys` is a string array that is used for grouping the data. If it is empty, the processor performs compaction and reassociates all spans with matching Resource and InstrumentationLibrary.
+
+
+### output block
+
+{{< docs/shared lookup="reference/components/output-block.md" source="alloy" version="<ALLOY_VERSION>" >}}
+
+### debug_metrics block
+
+{{< docs/shared lookup="reference/components/otelcol-debug-metrics-block.md" source="alloy" version="<ALLOY_VERSION>" >}}
+
+## Exported fields
+
+The following fields are exported and can be referenced by other components:
+
+| Name    | Type               | Description                                                   |
+|---------|--------------------|---------------------------------------------------------------|
+| `input` | `otelcol.Consumer` | Accepts `otelcol.Consumer` data for metrics, logs, or traces. |
+
+`input` accepts `otelcol.Consumer` data for any telemetry signal (metrics,
+logs, or traces).
+
+## Component health
+
+`otelcol.processor.groupbyattrs` is only reported as unhealthy if given an invalid
+configuration.
+
+## Debug information
+
+`otelcol.processor.groupbyattrs` doesn't expose any component-specific debug
+information.
+
+## Debug metrics
+
+`otelcol.processor.groupbyattrs` doesn't expose any component-specific debug metrics.
+
+## Examples
+
+### Grouping metrics by an attribute
+
+This example reassociates the metrics based on the value of the `host.name` attribute.
+
+```alloy
+otelcol.processor.groupbyattrs "default" {
+  keys = [
+    "host.name",
+  ]
+
+  output {
+    metrics = [otelcol.exporter.otlp.default.input]
+    logs    = [otelcol.exporter.otlp.default.input]
+    traces  = [otelcol.exporter.otlp.default.input]
+  }
+}
+```
+
+## Notes
+- The data points with different data types aren't merged under the same metric. For example, a gauge and sum metric would not be merged.
+- The data points without the specified keys remain under their respective resources.
+- New resources inherit the attributes of the original resource and the specified attributes in the keys array.
+- The grouping attributes in the keys array are removed from the output metrics.
```

Comment:
```suggestion
### Grouping metrics

Consider the below metrics, all originally associated to the same Resource:

``
Resource {host.name="localhost",source="prom"}
  Metric "gauge-1" (GAUGE)
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-B",id="eth0"}
  Metric "gauge-1" (GAUGE) // Identical to previous Metric
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-B",id="eth0"}
  Metric "mixed-type" (GAUGE)
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-B",id="eth0"}
  Metric "mixed-type" (SUM)
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-A",id="eth0"}
  Metric "dont-move" (Gauge)
    DataPoint {id="eth0"}
``

With the below configuration, the groupbyattrs will re-associate the metrics with either `host-A` or `host-B`, based on the value of the `host.name` attribute.

``alloy
otelcol.processor.groupbyattrs "default" {
  keys = [ "host.name" ]

  output {
    metrics = [otelcol.exporter.otlp.default.input]
  }
}
``

The output of the processor will therefore be:

``
Resource {host.name="localhost",source="prom"}
  Metric "dont-move" (Gauge)
    DataPoint {id="eth0"}

Resource {host.name="host-A",source="prom"}
  Metric "gauge-1"
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
  Metric "mixed-type" (GAUGE)
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
  Metric "mixed-type" (SUM)
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}

Resource {host.name="host-B",source="prom"}
  Metric "gauge-1"
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
  Metric "mixed-type" (GAUGE)
    DataPoint {id="eth0"}
``

This output demonstrates how `otelcol.processor.groupbyattrs` works in various situations:

- The DataPoints for the `gauge-1` (GAUGE) metric were originally split under 2 Metric instances and have been merged in the output.
- The DataPoints of the `mixed-type` (GAUGE) and `mixed-type` (SUM) metrics have not been merged under the same Metric, because their DataType is different.
- The `dont-move` metric DataPoints don't have a `host.name` attribute and therefore remained under the original Resource.
- The new Resources inherited the attributes from the original Resource (`source="prom"`), plus the specified attributes from the processed metrics (`host.name="host-A"` or `host.name="host-B"`).
- The specified "grouping" attributes that are set on the new Resources are also removed from the metric DataPoints.
- While not shown in the above example, the processor also merges collections of records under matching InstrumentationLibrary.
```

@clayton-cornell does this look ok? I copied it from the [upstream docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.105.0/processor/groupbyattrsprocessor/README.md#grouping-metrics).

@kehindesalaam Note that you'd need to replace the `` with ```. I can't do it in the GitHub suggestion, because I can't escape the tick symbol inside suggestions.

Reply:
@ptodev You can use codefence markdown in suggestions.. just use 4 backticks to wrap the suggestion block and the standard 3 backticks for code bloack.

Instead of `the below`, we can use `the following` (better grammatically).  I'll look at the info-flow a little deeper once it's applied and do another pass on the docs.

We have to be careful with 1:1 copied text from upstream - we should either directly link to it or note it's source (per whatever license upstream is using) if we copy it.

Reply:
I'll make the modification and also the adjustment Clayton suggested.